File: backend/app/schemas/vision.py
"""
Vision Analysis Schemas (Pydantic)

Request/response contracts for the deep learning vision endpoint.
"""

from typing import Optional

from pydantic import BaseModel, Field


class VisionAnalyzeRequest(BaseModel):
    """Request to analyze an image or PDF dataset with deep learning models."""
    dataset_id: int
    modes: list[str] = Field(
        default=["all"],
        description='Analysis modes to run. Options: "classify", "detect", "ocr", "all"',
    )
    top_k: int = Field(default=5, ge=1, le=20, description="Top-K classification predictions")
    confidence_threshold: float = Field(
        default=0.5, ge=0.1, le=1.0,
        description="Minimum confidence for object detection results",
    )


class ClassificationPrediction(BaseModel):
    label: str
    confidence: float


class ClassificationResult(BaseModel):
    model: str
    predictions: list[ClassificationPrediction]


class DetectionItem(BaseModel):
    label: str
    confidence: float
    bbox: list[float]  # [x1, y1, x2, y2]


class ImageSize(BaseModel):
    width: int
    height: int


class DetectionResult(BaseModel):
    model: str
    detections: list[DetectionItem]
    image_size: ImageSize


class OCRResult(BaseModel):
    method: str
    text: str
    char_count: int
    mean_confidence: Optional[float] = None
    word_count: Optional[int] = None


class VisionAnalyzeResponse(BaseModel):
    """
    Response containing results from requested vision analysis modes.
    Only the modes that were actually run will have non-null values.
    """
    dataset_id: int
    dataset_name: str
    file_type: str
    classification: Optional[ClassificationResult] = None
    detection: Optional[DetectionResult] = None
    ocr: Optional[OCRResult] = None

File: frontend/src/services/vision.ts
/**
 * Vision / Deep Learning API Service
 *
 * Runs pretrained model inference on image and PDF datasets:
 * - Image classification (ResNet18, ImageNet)
 * - Object detection (Faster R-CNN, COCO)
 * - OCR text extraction (Tesseract)
 */

import api from './api';

// ─── Types ────────────────────────────────────────────────────

export interface ClassificationPrediction {
  label: string;
  confidence: number;
}

export interface DetectionItem {
  label: string;
  confidence: number;
  bbox: [number, number, number, number]; // [x1, y1, x2, y2]
}

export interface VisionAnalysisResult {
  dataset_id: number;
  dataset_name: string;
  file_type: string;
  classification?: {
    model: string;
    predictions: ClassificationPrediction[];
  } | null;
  detection?: {
    model: string;
    detections: DetectionItem[];
    image_size: { width: number; height: number };
  } | null;
  ocr?: {
    method: string;
    text: string;
    char_count: number;
    mean_confidence?: number | null;
    word_count?: number | null;
  } | null;
}

// ─── API Function ─────────────────────────────────────────────

export async function analyzeDataset(
  datasetId: number,
  modes: string[] = ['all'],
  topK = 5,
  confidenceThreshold = 0.5
): Promise<VisionAnalysisResult> {
  const response = await api.post('/vision/analyze', {
    dataset_id: datasetId,
    modes,
    top_k: topK,
    confidence_threshold: confidenceThreshold,
  });
  return response.data;
}
